### ISSUE TITLE
Provide replay-safe ILogger in orchestrations

### ISSUE THREAD

**Issue by ulvesked** [microsoft/durabletask-dotnet#46]
At the moment, the orchestrator is created as a singleton using parameterless constructor. 

```cs
static readonly ITaskOrchestrator singletonOrchestrator = new DurablePocoPayloadSample.Orchestrator();
```

Would it be possible to create it using ActivatorUtilities.CreateInstance instead so we could inject ILogger, IOptions and the like?

**Comment by jviau** [comment]
DI is intentionally not supported on orchestrations as it typically is _not_ what you want to do. For example, the two services you mentioned: injecting ILogger and IOptions can lead to issues for two reasons:

ILogger: you want a replay-safe logger to avoid duplicate logs during replay. `TaskOrchestrationContext` has a way to wrap a logger in a replay safe logger, but I think there may be some work to make logger more useful (for example, we could just always provider a logger to orchestrators).
IOptions: this can easily lead to non-deterministic issues, as your IOptions values may change between replays, and affect downstream logic. `IOptionsSnapshot` won't work either, as replays would be in different service scopes (so snapshot will still change). It is recommended you use an activity to read options, so they will be locked in place after that single read. Maybe we could offer a convenient way to read options, but this is not too high of priority as it is not blocked as is.

Basically, any service you inject to an orchestrator would need to be aware of the orchestrator idempotence requirements and adhere to them. This would make increase the complexity of those services considerably. It is recommended to instead get whatever you need from a service via an Activity. The one exception is logging or metrics - which I think we can do some work to improve the experience for.

**Comment by ulvesked** [comment]
That makes sense. It was a logger I was after. I found the `CreateReplayLogger` method but I have no ILogger to pass to it.

**Comment by jviau** [comment]
@ulvesked you are right, I don't see a convenient way for you to get an `ILogger` in an orchestration. Let's repurpose this issue to track that specific request.

**Comment by Tyler-V** [comment]
`_log = taskOrchestrationContext.CreateReplaySafeLogger(functionContext.GetLogger(nameof(MyOrchestratorName)));`

This appeared to work for me but it is quite a mouthful, providing `ILogger logger` on `RunAsync` was nice for in-process but resolving this in DI makes sense too
